cpp/src/datetime/datetime_ops.cu

```diff
@@ -0,0 +1,175 @@
+
+      switch (DatetimeComponent) {
+        case cudf::datetime::detail::year: return static_cast<int>(year_month_day(days_since_epoch).year());
+        case cudf::datetime::detail::month: return static_cast<unsigned>(year_month_day(days_since_epoch).month());
+        case cudf::datetime::detail::day: return static_cast<unsigned>(year_month_day(days_since_epoch).day());
+        case cudf::datetime::detail::weekday: return year_month_weekday(days_since_epoch).weekday().iso_encoding();
+        default: break;
+      }
+
+      auto time_since_midnight = ts - days_since_epoch;
+
+      if (time_since_midnight.count() < 0) {
+        time_since_midnight += days(1);
+      }
+
+      auto hrs_ = duration_cast<hours>(time_since_midnight);
+      auto mins_ = duration_cast<minutes>(time_since_midnight - hrs_);
+      auto secs_ = duration_cast<seconds>(time_since_midnight - hrs_ - mins_);
+
+      switch (DatetimeComponent) {
+        case cudf::datetime::detail::hour: return hrs_.count();
+        case cudf::datetime::detail::minute: return mins_.count();
+        case cudf::datetime::detail::second: return secs_.count();
+        default: return 0;
+      }
+    }
+  };
+
+  template <datetime_component DatetimeComponent>
+  struct launch_extract_component {
+
+    column_device_view input;
+    mutable_column_view output;
+
+    launch_extract_component(column_device_view inp, mutable_column_view out) : input(inp), output(out) {}
+
+    template <typename Element>
+    typename std::enable_if_t<not cudf::is_timestamp_t<Element>::value, void>
+    operator()(cudaStream_t stream) {
+      CUDF_FAIL("Cannot extract datetime component from non-timestamp column.");
+    }
+
+    template <typename Element>
+    typename std::enable_if_t<cudf::is_timestamp_t<Element>::value, void>
+    operator()(cudaStream_t stream) {
+      auto functor = extract_component_operator<DatetimeComponent>{input};
+      auto bound_f = [=] __device__ (size_type const i) {
+        return functor.template operator()<Element>(i);
+      };
+      thrust::tabulate(rmm::exec_policy(stream)->on(stream),
+                       output.begin<int16_t>(),
+                       output.end<int16_t>(),
+                       bound_f);
+    }
+  };
+
+  template <datetime_component DatetimeComponent>
+  std::unique_ptr<column> extract_component(column_view const& input, cudaStream_t stream) {
+
+    auto null_mask_state = input.nullable() ? mask_state::UNINITIALIZED : mask_state::UNALLOCATED;
+    auto output = make_numeric_column(data_type{type_id::INT16},
+                                      input.size(), null_mask_state,
+                                      stream, rmm::mr::get_default_resource());
+
+    auto d_input = column_device_view::create(input);
+    auto m_output = static_cast<mutable_column_view>(*output);
+    auto launch = launch_extract_component<DatetimeComponent>{*d_input, m_output};
+
+    if (null_mask_state == mask_state::UNINITIALIZED) {
```

Comment:
This block copies the input column's `null_mask` if it has one. I couldn't use `allocate_like` because the input/output dtypes are different. Is there a better way to do this?

Reply:
Hm, I think this exposes an area where we need an additional `make_numeric_column` factory where you can give it a pre-existing null mask.

Something like this:
```
template <typename B>
unique_ptr<column> make_numeric_column(data_type type, size_type size, B&& null_mask, cudaStream_t stream, rmm::mr::device_memory_resource * mr)
```

This way you can pass in a `device_buffer` for the null mask either by l-value or r-value reference which will deep-copy/move the contents.